src/main/java/org/opensearch/knn/index/KNNCircuitBreakerUtil.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index;
+
+import lombok.extern.log4j.Log4j2;
+import org.opensearch.action.admin.cluster.settings.ClusterUpdateSettingsRequest;
+import org.opensearch.action.admin.cluster.settings.ClusterUpdateSettingsResponse;
+import org.opensearch.client.Client;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.common.unit.ByteSizeValue;
+import org.opensearch.knn.plugin.stats.StatNames;
+import org.opensearch.knn.plugin.transport.KNNStatsAction;
+import org.opensearch.knn.plugin.transport.KNNStatsNodeResponse;
+import org.opensearch.knn.plugin.transport.KNNStatsRequest;
+import org.opensearch.knn.plugin.transport.KNNStatsResponse;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+import static org.opensearch.knn.index.KNNCircuitBreaker.KNN_CIRCUIT_BREAKER_TRIGGERED;
+import static org.opensearch.knn.index.KNNCircuitBreaker.KNN_CIRCUIT_BREAKER_TRIGGERED_SETTING;
+import static org.opensearch.knn.index.KNNCircuitBreaker.KNN_CIRCUIT_BREAKER_UNSET_PERCENTAGE_SETTING;
+import static org.opensearch.knn.index.KNNCircuitBreaker.KNN_MEMORY_CIRCUIT_BREAKER_LIMIT_SETTING;
+
+/**
+ * Singleton utility class to interact with the circuit breaker.
+ */
+@Log4j2
+public class KNNCircuitBreakerUtil {
+    private static KNNCircuitBreakerUtil instance;
+
+    private KNNCircuitBreakerUtil() {}
+
+    private Client client;
+
+    /**
+     * Return instance of the KNNCircuitBreakerUtil, must be initialized first for proper usage
+     * @return instance of KNNCircuitBreakerUtil
+     */
+    public static synchronized KNNCircuitBreakerUtil instance() {
+        if (instance == null) {
+            instance = new KNNCircuitBreakerUtil();
+        }
+        return instance;
+    }
+
+    /**
+     *
+     * @param client client for interfacing with the cluster
+     */
+    public void initialize(final Client client) {
+        this.client = client;
+    }
+
+    /**
+     * Get the nodes in cluster that have reached the maximum capacity
+     *
+     * @return list of IDs for nodes that have reached the max capacity
+     */
+    public List<String> getNodesAtMaxCapacity() throws ExecutionException, InterruptedException {
+        KNNStatsRequest knnStatsRequest = new KNNStatsRequest();
+        knnStatsRequest.addStat(StatNames.CACHE_CAPACITY_REACHED.getName());
+        knnStatsRequest.timeout(new TimeValue(1000 * 10)); // 10 second timeout
+
+        KNNStatsResponse knnStatsResponse = client.execute(KNNStatsAction.INSTANCE, knnStatsRequest).get();
```

Comment:
since initialize is outside object creation, can you add non-null check for client?